File: java/patterns-use-cases/event-processing-enrichment/src/main/java/my/example/PackageTracker.java
package my.example;

import dev.restate.sdk.ObjectContext;
import dev.restate.sdk.SharedObjectContext;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Shared;
import dev.restate.sdk.annotation.VirtualObject;
import dev.restate.sdk.common.StateKey;
import dev.restate.sdk.common.TerminalException;
import dev.restate.sdk.http.vertx.RestateHttpEndpointBuilder;
import dev.restate.sdk.serde.jackson.JacksonSerdes;
import my.example.types.LocationUpdate;
import my.example.types.PackageInfo;

// Package tracking system:
// Digital twin representing a package in delivery with real-time location updates.
// Handlers get called over HTTP or Kafka.
@VirtualObject
public class PackageTracker {

    private static final StateKey<PackageInfo> PACKAGE_INFO =
            StateKey.of("package-info", JacksonSerdes.of(PackageInfo.class));

    // Called first by the seller over HTTP
    @Handler
    public void registerPackage(ObjectContext ctx, PackageInfo packageInfo){
        // Store the package details in the state
        ctx.set(PACKAGE_INFO, packageInfo);
    }

    // Connected to a Kafka topic for real-time location updates
    @Handler
    public void updateLocation(ObjectContext ctx, LocationUpdate locationUpdate){
        var packageInfo = ctx.get(PACKAGE_INFO)
                .orElseThrow(() -> new TerminalException("Package not found"));

        // Update the package info with the new location
        packageInfo.addLocation(locationUpdate);
        ctx.set(PACKAGE_INFO, packageInfo);
    }

    // Called by the delivery dashboard to get the package details
    @Shared
    public PackageInfo getPackageInfo(SharedObjectContext ctx){
        return ctx.get(PACKAGE_INFO)
                .orElseThrow(() -> new TerminalException("Package not found"));
    }

    public static void main(String[] args) {
        RestateHttpEndpointBuilder.builder()
                .bind(new PackageTracker())
                .buildAndListen(9081);
    }
}

// Example API Usage:
/*
curl localhost:8080/PackageTracker/package123/registerPackage -H 'content-type: application/json' -d '{ "finalDestination": "Bridge 6, Amsterdam"}'
curl localhost:8080/PackageTracker/package123/updateLocation -H 'content-type: application/json' -d '{ "timestamp": "2024-12-11T12:00:00Z", "location": "Warehouse A" }'
curl localhost:8080/PackageTracker/package123/getPackageInfo
*/

File: python/patterns-use-cases/src/eventenrichment/package_tracker.py
from typing import List

import restate
from pydantic import BaseModel
from restate import VirtualObject, ObjectContext
from restate.exceptions import TerminalError


class LocationUpdate(BaseModel):
    location: str
    timestamp: str


class PackageInfo(BaseModel):
    final_destination: str
    locations: List[LocationUpdate] = []


# Package tracking system:
# Digital twin representing a package in delivery with real-time location updates.
# Handlers get called over HTTP or Kafka.
package_tracker = VirtualObject("package-tracker")


# Called first by the seller over HTTP
@package_tracker.handler("registerPackage")
async def register_package(ctx: ObjectContext, package_info: PackageInfo):
    # store in state the user's information as coming from the registration event
    ctx.set("package-info", package_info.model_dump())


# Connected to a Kafka topic for real-time location updates
@package_tracker.handler("updateLocation")
async def update_location(ctx: ObjectContext, location_update: LocationUpdate):
    # get the package info from the state
    package_info = PackageInfo(**await ctx.get("package-info"))
    if package_info is None:
        raise TerminalError(f"Package {ctx.key()} not found")

    # Update the package details in the state
    locations = package_info.locations or []
    locations.append(location_update)
    package_info.locations = locations

    # store the updated package info in state
    ctx.set("package-info", package_info.model_dump())


# Called by the delivery dashboard to get the package details
@package_tracker.handler("getPackageInfo")
async def get_package_info(ctx: ObjectContext) -> PackageInfo:
    return PackageInfo(**await ctx.get("package-info"))


app = restate.app(services=[package_tracker])